src/ReactDocumentEvents.js

```diff
@@ -11,27 +12,36 @@ if (NODE_ENV !== 'production') {
 class DocumentEvents extends React.Component {
   // propTypes are generated below from all possible events
 
+  constructor(props) {
+    super(props);
+    this._allHandlers = {};
+  }
+
   componentDidMount() {
-    if (this.props.enabled) this.bindHandlers();
+    if (this.props.enabled) this.bindHandlers(this.props);
   }
 
   componentWillUnmount() {
-    this.unbindHandlers();
+    this.unbindHandlers(this.props);
   }
 
   componentWillReceiveProps(nextProps) {
-    if (this.props.enabled && !nextProps.enabled) {
-      this.unbindHandlers();
+    if (!shallowequal(Object.keys(this.props).sort(), Object.keys(nextProps).sort())) {
```

Comment:
Unclear on how this actually solves the issue with rebinding the handler. If you are passing, e.g. an `'onMouseMove'` function that changes from time to time because it's generated by a hook, this will not catch it.

I also don't think that `shallowequal()` is really needed here. You could just compare the arrays by string (`if (Object.keys(this.props).sort().toString() === Object.keys(nextProps).sort().toString())`).